src/backend/access/transam/xlog.c

```diff
@@ -9366,3 +9366,26 @@ SetWalWriterSleeping(bool sleeping)
 	XLogCtl->WalWriterSleeping = sleeping;
 	SpinLockRelease(&XLogCtl->info_lck);
 }
+
+void
+XLogUpdateWalBuffers(char* data, XLogRecPtr start, size_t len)
+{
+	XLogRecPtr end;
+	int idx;
+	XLogRecPtr pagebegptr;
+
+	LWLockAcquire(WALBufMappingLock, LW_EXCLUSIVE);
+
+	end = start + len;
+	idx = XLogRecPtrToBufIdx(end);
+	pagebegptr = XLogCtl->xlblocks[idx] - XLOG_BLCKSZ;
+
+	if (pagebegptr + XLOG_BLCKSZ >= end && pagebegptr < end)
+	{
+		/* Last page of the segment is present in WAL buffers */
+		char* page = &XLogCtl->pages[idx * XLOG_BLCKSZ];
+		size_t overlap = end - pagebegptr;
```

Comment:
How you can be sure chunk contains data since beginning of the page? Probably it doesn't, then memcpy src points to random place.

Reply:
Sorry, may be I do not understand you,  but `end` corresponds to LSN of segment end. We find out it offset from  LSN of start of the page. And copy data from this location. So `data + len - overlap` corresponds to `pagebegptr`. It is start of the page. And it doesn't matter where `buf` contains data since beginning of the page, does it?

Reply:
In your PR `XLogUpdateWalBuffers` is called from `XLogWalPropWrite`, which receives WAL stream from safekeeper. It means `data` points to WAL in [start; start + len) range, which is quite arbitrary. It might not include beginning of the page.

Reply:
Certainly it may not contain beginning of the page. But I this case we just do nothing for this chunk. Beginning of the page may be in another chunk. In this case we will copy it from this chunk. Function  `XLogUpdateWalBuffers` is called for all received chunks.

Reply:
Ok, I see a fix was pushed.